pkg/lib/hooks.js

```diff
@@ -0,0 +1,291 @@
+ * The "content" and "error" variables will be null until the file has
+ * been read for the first time.
+ *
+ * useFile and useFileWithError are pretty much the same. useFile will
+ * hide the exact error from the caller, which makes it slightly
+ * cleaner to use when the exact error is not part of the UI. In the
+ * case of error, useFile will log that error to the console.
+ */
+
+export function useFileWithError(path, options, hook_options) {
+    const [content_and_error, setContentAndError] = useState([null, null]);
+    const memo_options = useDeepEqualMemo(options);
+    const memo_hook_options = useDeepEqualMemo(hook_options);
+
+    useEffect(() => {
+        const handle = cockpit.file(path, memo_options);
+        handle.watch((data, tag, error) => {
+            setContentAndError([data || false, error || false]);
+            if (!data && memo_hook_options && memo_hook_options.log_errors)
+                console.warn("Can't read " + path + ": " + (error ? error.toString() : "not found"));
+        });
+        return function () {
+            handle.close();
+        };
+    }, [path, memo_options, memo_hook_options]);
+
+    return content_and_error;
+}
+
+export function useFile(path, options) {
+    const [content] = useFileWithError(path, options, { log_errors: true });
+    return content;
+}
+
+/* - useObject(create, destroy, deps)
+ *
+ * function Component(param) {
+ *   const obj = useObject(() => create_object(param),
+ *                         obj => obj.close(),
+ *                         [param], [deep_equal])
+ *
+ *   ...
+ * }
+ *
+ * This will call "create_object(param)" before the first render of
+ * the component, and will call "obj.close()" after the last render.
+ *
+ * More precisely, create_object will be called as part of the first
+ * call to useObject, i.e., at the very beginning of the first render.
+ *
+ * When "param" changes compared to the previous call to useObject
+ * (according to the deep_equal function in the example above), the
+ * object will also be destroyed and a new one will be created for the
+ * new value of "param" (as part of the call to useObject).
+ *
+ * There is no time when the "obj" variable is null in the example
+ * above; the first render already has a fully created object.  This
+ * is an advantage that useObject has over useEffect, which you might
+ * otherwise use to only create objects when dependencies have
+ * changed.
+ *
+ * And unlike useMemo, useObject will run a cleanup function when a
+ * component is removed.  Also unlike useMemo, useObject guarantees
+ * that it will not ignore the dependencies.
+ *
+ * The dependencies are compared with Object.is by default, but you
+ * can override that with custom comparators as shown in the example
+ * above.
+ */
+
+function deps_changed(old_deps, new_deps, comps) {
+    return (!old_deps || old_deps.length != new_deps.length ||
+            old_deps.findIndex((o, i) => !(comps[i] || Object.is)(o, new_deps[i])) >= 0);
```

Comment:
Ah, so `comps` is an array which matches a comparison function to the corresponding index of `param`, ok; this is a little unwieldy on creation; my first instinct was "let's instead pass a list of either a single param or a (param, compare_function) pair", but that wouldn't work  if param itself is a list. So, let's keep that structure, but document it more clearly.

Reply:
Yes, added more documentation.